include/deal.II/lac/petsc_snes.templates.h

```diff
@@ -0,0 +1,418 @@
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::NonlinearSolver(
+    const NonlinearSolverData &data,
+    const MPI_Comm &           mpi_comm)
+  {
+    AssertSNES(SNESCreate(mpi_comm, &snes));
+    AssertSNES(SNESSetApplicationContext(snes, this));
+    reinit(data);
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::~NonlinearSolver()
+  {
+    AssertSNES(SNESDestroy(&snes));
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  void
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::reinit()
+  {
+    AssertSNES(SNESReset(snes));
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  void
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::reinit(
+    const NonlinearSolverData &data)
+  {
+    reinit();
+
+    // Solver type
+    if (data.snestype.size())
+      AssertSNES(SNESSetType(snes, data.snestype.c_str()));
+
+    // Linesearch type
+    if (data.sneslstype.size())
+      {
+        SNESLineSearch linesearch;
+        AssertSNES(SNESGetLineSearch(snes, &linesearch));
+        AssertSNES(SNESLineSearchSetType(linesearch, data.sneslstype.c_str()));
+      }
+
+    // Options prefix
+    if (data.opts_prefix.size())
+      AssertSNES(SNESSetOptionsPrefix(snes, data.opts_prefix.c_str()));
+
+    // Solver tolerances
+    PetscReal atol = data.absolute_tolerance > 0.0 ?
+                       data.absolute_tolerance :
+                       static_cast<PetscReal>(PETSC_DEFAULT);
+    PetscReal rtol = data.relative_tolerance > 0.0 ?
+                       data.relative_tolerance :
+                       static_cast<PetscReal>(PETSC_DEFAULT);
+    PetscReal stol = data.step_tolerance > 0.0 ?
+                       data.step_tolerance :
+                       static_cast<PetscReal>(PETSC_DEFAULT);
+
+    // Maximum number of iterations and function evaluations.
+    PetscInt maxit =
+      data.max_it >= 0 ? data.max_it : static_cast<PetscInt>(PETSC_DEFAULT);
+    PetscInt maxfe =
+      data.max_fe >= 0 ? data.max_fe : static_cast<PetscInt>(PETSC_DEFAULT);
+    AssertSNES(SNESSetTolerances(snes, atol, rtol, stol, maxit, maxfe));
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  void
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::reinit_matrix(
+    PMatrixType &P)
+  {
+    this->A = nullptr;
+    this->P = &P;
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  void
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::reinit_matrices(
+    AMatrixType &A,
+    PMatrixType &P)
+  {
+    this->A = &A;
+    this->P = &P;
+  }
+
+
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  unsigned int
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::solve(VectorType &x)
+  {
+    auto snes_function_ =
```

Comment:
```suggestion
    const auto snes_function_ =
```

Reply:
From a naive C++ implementor: why is this so important?

Reply:
Ok, I will make `const` everything I can